### PR TITLE
 Make PInvokeStaticSigInfo always report errors

### PR DESCRIPTION
**Pull request by elinor-fung**
We had one usage of `PInvokeStaticSigInfo` that specified to not report errors. The only thing it actually wanted was the calling convention, not all the other information parsed by that class.

This creates a separate function specifically for getting only the calling convention for a p/invoke or delegate without throwing on errors, so that `PInvokeStaticSigInfo` does not need to always track errors and avoid throwing.

cc @AaronRobinsonMSFT @jkoritzinsky

### CODE REVIEW

File: src/coreclr/vm/dllimport.cpp [hunk 1]
```diff
@@ -2548,6 +2548,22 @@ namespace
...
     }
+
+    HRESULT ParseCallingConventionFromAttributeConstructor(_Inout_ CustomAttributeParser& ca, _Out_ CorInfoCallConvExtension* callConv)
+    {
+        LIMITED_METHOD_CONTRACT;
+        _ASSERTE(callConv != NULL);
+
+        CaArg callConvArg;
+        callConvArg.InitEnum(SERIALIZATION_TYPE_I4, (ULONG)0);
+        HRESULT hr = ParseKnownCaArgs(ca, &callConvArg, 1);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2559]: I believe this can return `S_FALSE`.

File: src/coreclr/vm/dllimport.cpp [hunk 2]
```diff
@@ -2809,14 +2807,10 @@ void PInvokeStaticSigInfo::DllImportInit(_In_ MethodDesc* pMD, _Outptr_opt_ LPCU
...
-    if (SUCCEEDED(RemapLinkType(nlt, &nlt)))
-    {
-        SetCharSet(nlt);
-    }
-    else
-    {
-        SetError(IDS_EE_NDIRECT_BADNATL);
-    }
+    if (FAILED(RemapLinkType(nlt, &nlt)))
+        ReportError(IDS_EE_NDIRECT_BADNATL);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2811]: Can we rename `ReportError` to `ThrowError`?

File: src/coreclr/vm/dllimport.cpp [hunk 3]
```diff
@@ -2994,25 +2951,82 @@ void PInvokeStaticSigInfo::InitCallConv(CorInfoCallConvExtension callConv, BOOL
...
 
     if (bIsVarArg && m_callConv != CorInfoCallConvExtension::C)
-        SetError(IDS_EE_NDIRECT_BADNATL_VARARGS_CALLCONV);
+        ReportError(IDS_EE_NDIRECT_BADNATL_VARARGS_CALLCONV);
 
     _ASSERTE(m_callConv != CallConvWinApiSentinel);
 }
 
-void PInvokeStaticSigInfo::ReportErrors()
+void PInvokeStaticSigInfo::ReportError(WORD error)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2959]: `error` is really "message ID" right?

File: src/coreclr/vm/dllimport.cpp [hunk 4]
```diff
@@ -2994,25 +2951,82 @@ void PInvokeStaticSigInfo::InitCallConv(CorInfoCallConvExtension callConv, BOOL
...
+{
+    CONTRACTL
+    {
+        STANDARD_VM_CHECK;
+        PRECONDITION(pMD != NULL);
+        PRECONDITION(pMD->IsNDirect());
+    }
+    CONTRACTL_END;
+
+    // This method explicitly does not check that any calling convention specified through
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2983]: ~~explicitly~~ -> intentionally?

File: src/coreclr/vm/dllimport.cpp [hunk 5]
```diff
@@ -2738,16 +2747,6 @@ void PInvokeStaticSigInfo::DllImportInit(_In_ MethodDesc* pMD, _Outptr_opt_ LPCU
...
     if (FAILED(pInternalImport->GetPinvokeMap(pMD->GetMemberDef(), (DWORD*)&mappingFlags, ppEntryPointName, &modref)))
     {
-#if !defined(CROSSGEN_COMPILE) // IJW
-        // The guessing heuristic has been broken with NGen for a long time since we stopped loading
-        // images at NGen time using full LoadLibrary. The DLL references are not resolved correctly
-        // without full LoadLibrary.
-        //
-        // Disable the heuristic consistently during NGen so that it does not kick in by accident.
-        if (!IsCompilationProcess())
-            BestGuessNDirectDefaults(pMD);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2748]: I'm missing where this was moved.

Comment (elinor-fung) [reply src/coreclr/vm/dllimport.cpp:2748]: It was just deleted. `BestGuessNDirectDefaults` was simply finding the would-be target of the call and then doing nothing with it.

It used to check if the target looked like a Win32 API call and set `nlfLastError` if it did. That stopped being necessary long ago - the VC linker handles it (in the Whidbey timeframe based on comments in history). The portion setting that flag was removed, but the lead-up to it wasn't.

File: src/coreclr/vm/dllimport.cpp (hunk 1 above)
Comment (elinor-fung) [reply src/coreclr/vm/dllimport.cpp:2559]: I don't see where?
All our existing usage also seems to just check for failed.

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/dllimport.cpp:2559]: Oh sorry. I was confused about the finding a Custom Attribute logic and parsing. You're correct.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**